File: src/pages/api/correlation.ts
// Section F6: Rolling Correlation
// Germany removed: IRLTLT01DEM156N is monthly. Aligning monthly with daily data
// produces ~24 matched dates over 2 years — far too sparse for 30d or 90d rolling windows.
// We now show only NO 10Y vs US 10Y (both daily, robust result).

import type { NextApiRequest, NextApiResponse } from "next";
import { fredSeries, nDaysAgo } from "@/lib/fred";
import { nbFetchYield } from "@/lib/norgesbank";
import { getCache, setCache } from "@/lib/cache";
import { alignByDate, rollingCorrelation } from "@/lib/stats";

const CACHE_KEY = "correlation";

export interface CorrelationPoint {
  date: string;
  noUsCorr: number | null;
}

export interface CorrelationResponse {
  data30d: CorrelationPoint[];
  data90d: CorrelationPoint[];
  updatedAt: string;
  note: string;
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<CorrelationResponse | { error: string }>
) {
  const cached = getCache<CorrelationResponse>(CACHE_KEY);
  if (cached) return res.status(200).json(cached);

  // Need 2 years + 90d window buffer
  const start = nDaysAgo(365 * 2 + 120);

  try {
    const [usRaw, noRaw] = await Promise.all([
      fredSeries("DGS10", start).catch(() => []),
      nbFetchYield("10Y", start).catch(() => []),
    ]);

    const us = usRaw.map((p) => ({ date: p.date, value: parseFloat(p.value) }));
    const no = noRaw.map((p) => ({ date: p.date, value: p.value }));

    // Align by date (both daily — robust alignment)
    const aligned = alignByDate(no, us);

    if (aligned.length < 30) {
      return res.status(200).json({
        data30d: [], data90d: [],
        updatedAt: new Date().toISOString(),
        note: "Insufficient data",
      });
    }

    const noVals = aligned.map((p) => p.a);
    const usVals = aligned.map((p) => p.b);

    const buildSeries = (window: number): CorrelationPoint[] => {
      const corrs = rollingCorrelation(noVals, usVals, window);
      return aligned.map((p, i) => ({
        date: p.date,
        noUsCorr: corrs[i] !== null ? parseFloat(corrs[i]!.toFixed(3)) : null,
      }));
    };

    const cutoff = nDaysAgo(365 * 2);
    const trim = (pts: CorrelationPoint[]) => pts.filter((p) => p.date >= cutoff);

    const result: CorrelationResponse = {
      data30d: trim(buildSeries(30)),
      data90d: trim(buildSeries(90)),
      updatedAt: new Date().toISOString(),
      note: "Germany excluded: monthly frequency (FRED IRLTLT01DEM156N) is incompatible with 30d/90d rolling windows on daily data.",
    };

    setCache(CACHE_KEY, result);
    res.status(200).json(result);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
}

File: backend/routers/correlation.py
from datetime import datetime

import asyncio
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from lib.cache import get_cache, set_cache
from lib.fred import fred_series, n_days_ago
from lib.norgesbank import nb_fetch_yield
from lib.stats import align_by_date, rolling_correlation

router = APIRouter()
CACHE_KEY = "correlation"


@router.get("/correlation")
async def correlation():
    cached = get_cache(CACHE_KEY)
    if cached:
        return cached

    # Need 2 years + 90d window buffer
    start = n_days_ago(365 * 2 + 120)

    try:
        us_raw, no_raw = await asyncio.gather(
            fred_series("DGS10", start),
            nb_fetch_yield("10Y", start),
            return_exceptions=True,
        )
        if isinstance(us_raw, Exception):
            us_raw = []
        if isinstance(no_raw, Exception):
            no_raw = []

        aligned = align_by_date(no_raw, us_raw)

        if len(aligned) < 30:
            return {
                "data30d": [],
                "data90d": [],
                "updatedAt": datetime.utcnow().isoformat() + "Z",
                "note": "Insufficient data",
            }

        no_vals = [p["a"] for p in aligned]
        us_vals = [p["b"] for p in aligned]

        def build_series(window: int) -> list[dict]:
            corrs = rolling_correlation(no_vals, us_vals, window)
            return [
                {
                    "date": p["date"],
                    "noUsCorr": round(corrs[i], 3) if corrs[i] is not None else None,
                }
                for i, p in enumerate(aligned)
            ]

        cutoff = n_days_ago(365 * 2)

        def trim(pts: list[dict]) -> list[dict]:
            return [p for p in pts if p["date"] >= cutoff]

        result = {
            "data30d": trim(build_series(30)),
            "data90d": trim(build_series(90)),
            "updatedAt": datetime.utcnow().isoformat() + "Z",
            "note": (
                "Germany excluded: monthly frequency (FRED IRLTLT01DEM156N) is "
                "incompatible with 30d/90d rolling windows on daily data."
            ),
        }
        set_cache(CACHE_KEY, result)
        return result
    except Exception as e:
        return JSONResponse(status_code=500, content={"error": str(e)})
